### <b> Course: Machine Learning for Absolute Beginners - Level 2
#### By: Idan Gabrieli
####  Exercise #3 – Data Loading and Analysis
*****************************

### Step 1 - Loading a Tabular Data File

Make sure the "titanic.csv" file (provided as part of this excercise) is located in the JupyterLab working directory. <br> Read the "titanic.csv" file dataset into a dataframe called df. Print the first 5 rows in df.

In [1]:
import pandas as pd
df = pd.read_csv("titanic.csv")
df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket". Print the first 5 rows in df.

In [2]:
df = pd.read_csv("titanic.csv",usecols=[0,1,2,3,4,5,9,10,11])
df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,8.0500,NaN,S


Print the data type of each column by using the dtypes attribute.

In [3]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket" and also convert the column "Survived" into a boolean data type. Print the first 5 rows in df.

In [4]:
df = pd.read_csv("titanic.csv",usecols=[0,1,2,3,4,5,9,10,11],dtype={"Survived":"bool"})
df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
0,1,False,3,"Braund, Mr. Owen Harris",male,22.0,7.2500,NaN,S
1,2,True,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,71.2833,C85,C
2,3,True,3,"Heikkinen, Miss. Laina",female,26.0,7.9250,NaN,S
3,4,True,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,53.1000,C123,S
4,5,False,3,"Allen, Mr. William Henry",male,35.0,8.0500,NaN,S


Print the data type of each column by using the dtypes attribute. The "Survived" should be boolean data type.

In [5]:
df.dtypes

PassengerId      int64
Survived          bool
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

### Step 2 - Preview the DataFrame

Print the two index components of the df dataframe: index, columns.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    bool   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   Fare         891 non-null    float64
 7   Cabin        204 non-null    object 
 8   Embarked     889 non-null    object 
dtypes: bool(1), float64(2), int64(2), object(4)
memory usage: 56.7+ KB


Print the size of the dataframe using the shape attribute. We have 891 rows and 9 columns. 

In [7]:
df.shape

(891, 9)

Print the amount of rows per each column without missing values.

In [8]:
df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
Fare           891
Cabin          204
Embarked       889
dtype: int64

Print the unique values for the "Sex" column.

In [9]:
df["Sex"].unique()

array(['male', 'female'], dtype=object)

Print the unique values for the "Embarked" column.

In [10]:
df["Embarked"].unique()

array(['S', 'C', 'Q', nan], dtype=object)

Print the amount of unique values for the "Embarked" column.

In [11]:
df["Embarked"].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

Print the number of rows with the same specific value in the "Embarked" column.

In [12]:
df[df["Embarked"]=="S"]["Embarked"].count()
 

644

How many passengers survived?

In [13]:
df[df["Survived"]==True]["Survived"].count()
#Alternate: df['Survived'].value_counts()

342

What is the survival percentage?

In [14]:
df[df["Survived"]==True]["Survived"].count()/df["Survived"].count() *100
#Alternate: df['Survived'].value_counts(normalize=True)

38.38383838383838

### Step 3 - Using Summary Statistics

What is the maximum age of the titanic passengers?

In [15]:
df['Age'].max()

80.0

What is the average age and average ticket price?

In [16]:
print(df['Age'].mean())
df['Fare'].mean()

29.69911764705882


32.2042079685746

How many passengers survived? (using the sum method).

In [17]:
df['Survived'].sum()

342

Print a full statistical summary using the describe method. What is the agev value that 25% of passengers are smaller then this age? (should be 20.123).

In [18]:
df.describe()

,PassengerId,Pclass,Age,Fare
count,891.000000,891.000000,714.000000,891.000000
mean,446.000000,2.308642,29.699118,32.204208
std,257.353842,0.836071,14.526497,49.693429
min,1.000000,1.000000,0.420000,0.000000
25%,223.500000,2.000000,20.125000,7.910400
50%,446.000000,3.000000,28.000000,14.454200
75%,668.500000,3.000000,38.000000,31.000000
max,891.000000,3.000000,80.000000,512.329200


### Step 4 - Sorting and Ranking

Sort all rows in the dataframe using the 'Age' column is descending order and print the first 5 rows. 

In [19]:
df.sort_values('Age',ascending=False).head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
630,631,True,1,"Barkworth, Mr. Algernon Henry Wilson",male,80.0,30.0000,A23,S
851,852,False,3,"Svensson, Mr. Johan",male,74.0,7.7750,NaN,S
493,494,False,1,"Artagaveytia, Mr. Ramon",male,71.0,49.5042,NaN,C
96,97,False,1,"Goldschmidt, Mr. George B",male,71.0,34.6542,A5,C
116,117,False,3,"Connors, Mr. Patrick",male,70.5,7.7500,NaN,Q


Sort all rows in the dataframe using the 'Age' column is descending order and save it in the same dataframe. Print the first 5 rows. 

In [20]:
df = df.sort_values('Age',ascending=False)
df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
630,631,True,1,"Barkworth, Mr. Algernon Henry Wilson",male,80.0,30.0000,A23,S
851,852,False,3,"Svensson, Mr. Johan",male,74.0,7.7750,NaN,S
493,494,False,1,"Artagaveytia, Mr. Ramon",male,71.0,49.5042,NaN,C
96,97,False,1,"Goldschmidt, Mr. George B",male,71.0,34.6542,A5,C
116,117,False,3,"Connors, Mr. Patrick",male,70.5,7.7500,NaN,Q


Sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [21]:
df.sort_values(['Sex','Age'],ascending=[True,True])[['Sex','Name','Age']]

,Sex,Name,Age
644,female,"Baclini, Miss. Eugenie",0.75
469,female,"Baclini, Miss. Helene Barbara",0.75
381,female,"Nakid, Miss. Maria (""Mary"")",1.00
172,female,"Johnson, Miss. Eleanor Ileen",1.00
642,female,"Skoog, Miss. Margit Elizabeth",2.00
...,...,...,...
839,male,"Marechal, Mr. Pierre",NaN
846,male,"Sage, Mr. Douglas Bullen",NaN
859,male,"Razi, Mr. Raihed",NaN
868,male,"van Melkebeke, Mr. Philemon",NaN


Print the row in location 2 using the iloc method.

Drop the missing values and then sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [22]:
df.dropna().sort_values(['Sex', 'Age'], ascending=[True, True])[[ 'Sex', 'Name', 'Age']]

,Sex,Name,Age
205,female,"Strom, Miss. Telma Matilda",2.0
297,female,"Allison, Miss. Helen Loraine",2.0
10,female,"Sandstrom, Miss. Marguerite Rut",4.0
618,female,"Becker, Miss. Marion Louise",4.0
435,female,"Carter, Miss. Lucile Polk",14.0
...,...,...,...
54,male,"Ostby, Mr. Engelhart Cornelius",65.0
456,male,"Millet, Mr. Francis Davis",65.0
745,male,"Crosby, Capt. Edward Gifford",70.0
96,male,"Goldschmidt, Mr. George B",71.0


What are the five oldest passengers that paid a ticket that cost more than the average ticket price? (create a boolean condition and than use it to filter the dataframe and than use the nlargest.

In [23]:
df[df['Fare']>df['Fare'].mean()].nlargest(5,'Age')

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
493,494,False,1,"Artagaveytia, Mr. Ramon",male,71.0,49.5042,NaN,C
96,97,False,1,"Goldschmidt, Mr. George B",male,71.0,34.6542,A5,C
745,746,False,1,"Crosby, Capt. Edward Gifford",male,70.0,71.0000,B22,S
54,55,False,1,"Ostby, Mr. Engelhart Cornelius",male,65.0,61.9792,B30,C
438,439,False,1,"Fortune, Mr. Mark",male,64.0,263.0000,C23 C25 C27,S


### Step 5 - Filtering

Filter and keep only the male passengers.

In [24]:
df[df['Sex']=='male']

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
630,631,True,1,"Barkworth, Mr. Algernon Henry Wilson",male,80.0,30.0000,A23,S
851,852,False,3,"Svensson, Mr. Johan",male,74.0,7.7750,NaN,S
493,494,False,1,"Artagaveytia, Mr. Ramon",male,71.0,49.5042,NaN,C
96,97,False,1,"Goldschmidt, Mr. George B",male,71.0,34.6542,A5,C
116,117,False,3,"Connors, Mr. Patrick",male,70.5,7.7500,NaN,Q
...,...,...,...,...,...,...,...,...,...
839,840,True,1,"Marechal, Mr. Pierre",male,NaN,29.7000,C47,C
846,847,False,3,"Sage, Mr. Douglas Bullen",male,NaN,69.5500,NaN,S
859,860,False,3,"Razi, Mr. Raihed",male,NaN,7.2292,NaN,C
868,869,False,3,"van Melkebeke, Mr. Philemon",male,NaN,9.5000,NaN,S


Filer out all passengers below the age of 40 and that are also 'male'. Print the ten first results.

In [27]:
df[(df['Sex']=='male') & (df['Age']<40)].head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,Fare,Cabin,Embarked
528,529,False,3,"Salonen, Mr. Johan Werner",male,39.0,7.9250,NaN,S
13,14,False,3,"Andersson, Mr. Anders Johan",male,39.0,31.2750,NaN,S
811,812,False,3,"Lester, Mr. James",male,39.0,24.1500,NaN,S
808,809,False,2,"Meyer, Mr. August",male,39.0,13.0000,NaN,S
400,401,True,3,"Niskanen, Mr. Juha",male,39.0,7.9250,NaN,S
806,807,False,1,"Andrews, Mr. Thomas Jr",male,39.0,0.0000,A36,S
795,796,False,2,"Otter, Mr. Richard",male,39.0,13.0000,NaN,S
705,706,False,2,"Morley, Mr. Henry Samuel (""Mr Henry Marshall"")",male,39.0,26.0000,NaN,S
471,472,False,3,"Cacic, Mr. Luka",male,38.0,8.6625,NaN,S
465,466,False,3,"Goncalves, Mr. Manuel Estanslas",male,38.0,7.0500,NaN,S


### Step 6 - Grouping and Aggregating

What is the average ticket price for male versus female passengers?

In [29]:
df[["Sex", "Fare"]].groupby("Sex").mean()

,Fare
Sex,
female,44.479818
male,25.523893


What is the average age for the survived group?

In [31]:
df.groupby('Survived').mean()

,PassengerId,Pclass,Age,Fare
Survived,,,,
False,447.016393,2.531876,30.626179,22.117887
True,444.368421,1.950292,28.343690,48.395408


How many survived group by the embarked port?

In [35]:
df[['Embarked','Survived']].groupby('Embarked').sum()

,Survived
Embarked,
C,93
Q,30
S,217


What is the average age as well as the average price for male and female?

In [36]:
df[['Sex','Age','Fare']].groupby('Sex').mean()

,Age,Fare
Sex,,
female,27.915709,44.479818
male,30.726645,25.523893


### End of document
************************